builder/main.py

```diff
@@ -119,16 +219,37 @@ def generate_uf2(target, source, env):
     target_firm = join("$BUILD_DIR", "${PROGNAME}.bin")
 else:
     target_elf = env.BuildProgram()
-    target_firm = env.ElfToBin(join("$BUILD_DIR", "${PROGNAME}"), target_elf)
-    env.Depends(target_firm, "checkprogsize")
-
+    if set(["buildfs", "uploadfs"]) & set(COMMAND_LINE_TARGETS):
+        target_firm = env.DataToBin(
+            join("$BUILD_DIR", "${PICO_FS_IMAGE_NAME}"), "$PROJECTDATA_DIR")
+        AlwaysBuild(target_firm)
+    else:
+        target_firm = env.ElfToBin(join("$BUILD_DIR", "${PROGNAME}"), target_elf)
+        env.Depends(target_firm, "checkprogsize")
+
+env.AddPlatformTarget("buildfs", target_firm, target_firm, "Build Filesystem Image")
 AlwaysBuild(env.Alias("nobuild", target_firm))
 target_buildprog = env.Alias("buildprog", target_firm, target_firm)
 
 env.AddPostAction(
     target_elf, env.VerboseAction(generate_uf2, "Generating UF2 image")
 )
 
+def _update_max_upload_size(env):
+    fetch_fs_size(env)
+    env.BoardConfig().update("upload.maximum_size", env["PICO_FLASH_LENGTH"])
+
+# update max upload size based on set sketch size (or raw maximum size)
+if env.get("PIOMAINPROG"):
+    env.AddPreAction(
+        "checkprogsize",
+        env.VerboseAction(
+            lambda source, target, env: _update_max_upload_size(env),
+            "Retrieving maximum program size $SOURCE"))
+# remove after PIO Core 3.6 release
```

Comment:
PIO core 3.6 was released a long time ago?

Reply:
I believe so.

Reply:
Block of code was removed, upload size checks still work properly.